File: app/src/main/java/de/bikebean/app/db/settings/settings/replace_if_newer_settings/WarningNumber.java
package de.bikebean.app.db.settings.settings.replace_if_newer_settings;

import androidx.annotation.NonNull;

import de.bikebean.app.db.settings.settings.ReplaceIfNewerSetting;
import de.bikebean.app.db.sms.Sms;
import de.bikebean.app.db.state.State;
import de.bikebean.app.db.state.StateFactory;

public class WarningNumber extends ReplaceIfNewerSetting {

    private static final @NonNull State.KEY key = State.KEY.WARNING_NUMBER;

    private WarningNumber(final @NonNull Sms sms, final @NonNull String warningNumber,
                          final @NonNull State.STATUS status) {
        super(sms, StateFactory.createStringState(sms, key, warningNumber, status));
    }

    public WarningNumber(final @NonNull Sms sms,
                         final @NonNull WarningNumberGetter warningNumberGetter) {
        this(sms, warningNumberGetter.getWarningNumber(), State.STATUS.CONFIRMED);
    }

    public WarningNumber(final @NonNull Sms sms, final @NonNull String warningNumber) {
        this(sms, warningNumber, State.STATUS.UNSET);
    }

    public interface WarningNumberGetter {
        @NonNull String getWarningNumber();
    }
}

File: app/src/main/java/de/bikebean/app/db/settings/settings/replace_if_newer_settings/WarningNumber.kt
package de.bikebean.app.db.settings.settings.replace_if_newer_settings

import de.bikebean.app.db.settings.settings.ReplaceIfNewerSetting
import de.bikebean.app.db.sms.Sms
import de.bikebean.app.db.sms.SmsFactory
import de.bikebean.app.db.state.State
import de.bikebean.app.db.state.State.KEY
import de.bikebean.app.db.state.StateFactory

class WarningNumber private constructor(
        sms: Sms,
        warningNumber: String,
        status: State.STATUS
) : ReplaceIfNewerSetting(
        StateFactory.createStringState(sms, key, warningNumber, status)
) {

    constructor() : this(SmsFactory.createNullSms())

    constructor(sms: Sms) : this(sms, "", State.STATUS.UNSET)

    constructor(sms: Sms, warningNumber: String) : this(sms, warningNumber, State.STATUS.CONFIRMED)

    companion object {
        private val key = KEY.WARNING_NUMBER
    }

}